### PR TITLE
Support {gather} and {gather*}

### PR DESCRIPTION
**Pull request by ronkok**
This is the first in a series of pull requests aimed at support of AMS environments with automatic equation numbering. This PR supports `gather` and `gather*`. It establishes some methods for how to go about this sort of thing.

After (if?) this PR is approved and merged, subsequent PRs will address environments `equation`, `equation*`,  `align`, `align*`, `alignat`, and `alignat*`, `split`, `multline`, and `multline*`.

You can see both the HTML and the MathML if you use Firefox to view a [demo page](http://webmathcomparison.net/#g).

### Top Level Environment

In LaTeX, `{gather}` is itself a way to enter math mode. It will not render if called from math mode.  But KaTeX begins execution already in math mode. To reconcile the two, this PR adds a new rendering option, `topEnv`.  This enables an auto-render enhancement as suggested by @edemaine in #2042. The extension would pattern-match `\begin{gather}…\end{gather} ` and invoke KaTeX with the code and with rendering options `topEnv` and `displayMode` both set to true.

I expect that Markdown lexers will always use `$$…$$` and `$…$ ` to detect LaTeX. To support them, this PR also renders `{gather}` if invoked from display mode while `strict` is false or “warn”. Calling `{gather}` from inline mode always results in a KaTeX parse error.

The auto-render enhancement is not part of this PR. That’s for some other PR. And if anyone else wants to write it, feel free.

### CSS Counter

Automatic numbering is provided via CSS counters. There are two parallel counters, one for HTML and one for MathML.

CSS counters have several good features:

* They work in either client-side or server-side rendering.

* They can be augmented by an HTML page’s own CSS to display chapter numbers or section numbers along with the equation number.

* They evade all the difficulties that @kevinbarabash describes [here](https://github.com/KaTeX/KaTeX/issues/2121#issuecomment-548973467). 

CSS counters also have a couple of bad parts.

1. Multiple CSS counters need careful coordination. KaTeX CSS has a line that reads: `body { counter-reset: katexEqnNo mmlEqnNo; }`. Any subsequent counter-reset declaration will overwrite the KaTeX resets. So if an HTML page has its own CSS counter(s) they would have to write their counter-reset declaration to include both their own counter and the KaTeX counters, separated by spaces,.

2. A CSS counter cannot be the target of a link. But that can be fixed by a run-time extension. Say that a page is rendered sever-side. Here’s run time code that adds HTML id’s:

```js
const eqns = document.getElementsByClassName("eqn-num");
for (let i = 0; i < eqns.length; i++) {
    eqns[i].id = "eqn-num-" + String(i + 1);
}
```

Extensions could also be written to support  `\ref`, `\eqref`, and `\label`. Core KaTeX could write a span for each with its own class. Then a run-time extension could make two passes through the document. The first pass would add HTML ids to each counter, and record any `\label`s. The second pass would add an anchor element to each `\ref` that targets the appropriate id.

React fields and Vue fields would probably need a different sort of extension, if they support `\ref` at all. That extension is outside my expertise.

### To Do

Among the items left for future PR’s are:

- [ ] `equation`, `equation*`

- [ ] `split`

- [ ] `multline` and `multline*`

- [ ] `align`, `align*`, `alignat`, `alignat*`

- [ ] Upgrade of the auto-render extension

- [ ] `\ref`, `\eqref`, and `\label`

- [ ] `\notag`, `\shoveleft`, `\shoveright`

- [ ] Maybe someday `CD`

### CODE REVIEW

**Comment by ronkok** [comment]
And yes, I know that flow has an objection. I have no idea what it is talking about.

**Comment by ronkok** [comment]
Well, I fixed the typo but CircleCI is still running flow on the code from the previous commit and showing a flow error that should no longer exist. Odd.

**Comment by kevinbarabash** [comment]
This is great.  Restricting text-only environments to display mode makes a lot of sense.  As for the flow issue, the tests seems to be passing so it must have resolved itself.

**Comment by edemaine** [comment]
Wow, I just took my first look at the description of this PR, and it's really amazing! I didn't realize there was already a draft implementation of `topEnv` *and* equation numbering. Those are probably the most requested KaTeX features I've heard!  I'll try to review this one as well (though not today -- quickly running out of time).

**Comment by ronkok** [comment]
You know, I don't think we should rush this one. In this PR, I did the minimum that could get one of the environments to work. My plan was to finish the others after the general strategy was approved. I think it would be a good idea to do all the auto-numbering environments in the same release. That way, I could be more certain to avoid breaking changes later when we do `equation` and `split`.

**Comment by edemaine** [comment]
Yep, agreed. I'll still try to review it, but not before 0.12.0 release. 🙂

**Comment by danielhollas** [comment]
Is the demo page referenced in the OP running the latest commit? Looks like positioning is off
obrazek

In any case, once this gets merged I'd be happy to help write/test the modification for autorenderer. :-)

**Comment by ronkok** [comment]
> Looks like positioning is off

Could you be more specific?

**Comment by danielhollas** [comment]
Sure. The generated labels are overlapping with the equations in the KaTeX column on the screenshot above.
(tested in latest Firefox and Chromium on Linux)

**Comment by ronkok** [comment]
>  generated labels are overlapping with the equations in the KaTeX column

So they are, but this PR generates the KaTeX *MathML* column. The overlap you point out is rendered in the standard KaTeX HTML output. It's a valid comment, but it belongs in an issue to itself.

**Comment by ronkok** [comment]
@danielhollas Strike my last comment. I had something else on my mind. What I should have said is that the overlap occurs in both the proposed `{gather}`  environment and the pre-existing `tag` function. I suspect that the overlap can be eliminated with some better CSS, but I wanted to fix one problem at a time. This PR already has enough moving parts without trying to fix pre-existing problems on the fly.

**Comment by ronkok** [comment]
Note to reviewer: This is to explain why I (1) changed the treatment of `leqno` and (2) changed the MathML nested map functions into a for-loop. The pre-existing treatment of `leqno` happens at a high level in the KaTeX code.

In order to automatically place a tag node in the proper position, I needed access to `leqno`  within the `htmlBuilder` and  `mathmlBuilder` functions of `array.js`.  So I had to first change the treatment of `leqno` and attach it to the group that is passed to `htmlBuilder` and  `mathmlBuilder`. I unwound the MathML nested map functions for the same reason. I needed access to `leqno` inside a deeply nested callback and I could think of no other way to get it.

**Comment by ylemkimon** [comment]
Sorry for the delayed review. I was thinking about alternative approaches to CSS counter, but I couldn't think of one and months have passed :cry: I'll try to review this as soon as possible.

File: src/mathMLTree.js [hunk 1]
```diff
@@ -104,6 +115,10 @@ export class MathNode implements MathDomNode {
             }
         }
 
+        if (this.classes.length > 0) {
+            markup += ` class ="${createClass(this.classes)}"`;
```
Comment (ylemkimon) [inline src/mathMLTree.js:119]:
```suggestion
            markup += ` class ="${utils.escape(createClass(this.classes))}"`;
```

File: src/environments/array.js [hunk 2]
```diff
@@ -817,17 +868,31 @@ defineEnvironment({
...
     props: {
         numArgs: 0,
     },
     handler(context) {
+        if (utils.contains(["gather", "gather*"], context.envName)) {
+            const settings = context.parser.settings;
+            if (!settings.displayMode) {
+                throw new ParseError(`{${context.envName}} cannot be used inline.`);
+            } else if (settings.strict === "warn" && !settings.topEnv) {
+                // eslint-disable-next-line no-console
```
Comment (ylemkimon) [inline src/environments/array.js:881]:
```suggestion
```

File: src/environments/array.js [hunk 3]
```diff
@@ -817,17 +868,31 @@ defineEnvironment({
...
+                console.log(`{${context.envName}} called from math mode.`);
```
Comment (ylemkimon) [inline src/environments/array.js:882]:
```suggestion
                settings.reportNonstrict("textEnv",
                    `{${context.envName}} called from math mode.`);
```

File: src/environments/array.js [hunk 4]
```diff
@@ -817,17 +868,31 @@ defineEnvironment({
...
+    names: ["gathered", "gather", "gather*"],
     props: {
         numArgs: 0,
     },
     handler(context) {
+        if (utils.contains(["gather", "gather*"], context.envName)) {
+            const settings = context.parser.settings;
+            if (!settings.displayMode) {
+                throw new ParseError(`{${context.envName}} cannot be used inline.`);
+            } else if (settings.strict === "warn" && !settings.topEnv) {
```
Comment (ylemkimon) [inline src/environments/array.js:880]:
```suggestion
            } else if (settings.strict && !settings.topEnv) {
```

File: src/katex.less [hunk 5]
```diff
@@ -635,3 +635,23 @@
...
+}
+
+.mml-eqn-num::before {
+    counter-increment: mmlEqnNo;
+    content: "(" counter(mmlEqnNo) ")";
+}
+
+.mtr-glue {
+    width: 50%;
+}
```
Comment (ylemkimon) [inline src/katex.less:657]: I think these should be a `contrib` extension for now.

Comment (ronkok) [reply src/katex.less:657]: I agree that we should be careful with CSS counters. First, just to make sure that you and I understand each other, let me describe what I think is the potential problem. It lies in the CSS code `body { counter-reset: katexEqnNo mmlEqnNo; }`. The trouble is that a counter-reset statement can over-write all counters. This line of CSS would either over-write a page's existing counters or be over-written, depending on the order in which CSS files are loaded.

Now let me suggest another approach.

1. Remove from this PR the CSS line `body { counter-reset: katexEqnNo mmlEqnNo; }`.
2. Let the other CSS in this PR stand. It does no harm.
3. Amend the setup documentation to strongly suggest that web site administrators should include the `body { counter-reset: katexEqnNo mmlEqnNo; }` CSS in their page.
4. For convenience, create a `contrib` CSS file that has only one line, that being `body { counter-reset: katexEqnNo mmlEqnNo; }`.
5. Test this PR without a counter-reset and make sure that the worst outcome is the omission of automatic numbering and perhaps some console messages.

Am I addressing the same problem that you see? What do you think of my suggestion?

Comment (ylemkimon) [reply src/katex.less:657]: I was thinking moving all styles related to numbering out to a `contrib` module, but now I think it'd make KaTeX hard to configure. We could rather add an option to choose which numbering to use, e.g., CSS counter, Javascript, none, etc. (I'll send a PR later) I think `body { counter-reset: katexEqnNo mmlEqnNo; }` wouldn't cause too much trouble, so I think it's okay to leave it here.

Other styles, `.eqn-num::before`, `.mml-eqn-num::before`, and `.mtr-glue`, should be scoped, i.e., inside `.katex { ... }`.

**Comment by ylemkimon** [comment]
@danielhollas The overlap issue is tracked at #1983.

**Comment by ronkok** [comment]
In the most recent commit, I came across an odd behavior. Eslint now objects to any use of `console.error`, regardless of what sort of `eslint-disable` comment I apply. I even tried  `/*eslint-disable */` and `/*eslint-enable */`. In order to get a clean test run, I had to change `console.error` to `console.log`. This occurs in the `macro.js` terminal command `\errmessage`.

File: src/mathMLTree.js (hunk 1 above)
Comment (ronkok) [reply src/mathMLTree.js]: Done.

File: src/environments/array.js (hunk 3 above)
Comment (ronkok) [reply src/environments/array.js]: Done.

File: src/environments/array.js (hunk 4 above)
Comment (ronkok) [reply src/environments/array.js]: Done.

File: src/katex.less (hunk 5 above)
Comment (ronkok) [reply src/katex.less:657]: Agree 100% on scoping the styles inside `.katex{...}`. On it.

Comment (ronkok) [reply src/katex.less]: Styles are scoped inside `.katex{...}`.

If we let the CSS counter-reset remain as is, then I think the KaTeX setup documentation should be amended. That can perhaps be a later PR. I would like all the automatic numbering environments to go into the same KaTeX release.

**Comment by ylemkimon** [comment]
> In the error-handling commit, I came across an odd behavior. Eslint now objects to any use of `console.error`, regardless of what sort of `eslint-disable` comment I apply. I even tried  `/*eslint-disable */` and `/*eslint-enable */`. In order to get a clean test run, I had to change `console.error` to `console.log`. This occurs in the `macro.js` terminal command `\errmessage`.

It seems to work fine for me and [CircleCI](https://app.circleci.com/pipelines/github/KaTeX/KaTeX/2043/workflows/62715a0a-b4f3-4614-a07b-d764c083f29e/jobs/9721). Maybe wrong version of eslint got installed?

**Comment by ronkok** [comment]
> It seems to work fine for me and CircleCI. 

Okay, since it seems to be a local problem, I'll revert the `console.error` and continue my investigation locally.

**Comment by ronkok** [comment]
Yay! Thank you for the review. This one is pretty opinionated. There are decisions made and I'm glad you took the time to review them carefully.

**Comment by edemaine** [comment]
One question (maybe) before this merges: have you done a pros/cons analysis of using CSS counters vs. using the existing LaTeX macro system (extended to do LaTeX counters)?

**Comment by ronkok** [comment]
> pros/cons analysis of using CSS counters vs. using the existing LaTeX macro system (extended to do LaTeX counters)?

I did consider using some sort of Javascript counter. It would only have to carry an integer, so it could have been a lot simpler than the existing macro system. As you point out, this could be coordinated with the auto-render extension to provide a working system.

The problem that I could not overcome was the existence of fields in React, Vue, etc., that only update locally. A CSS counter handles this without drama. A Javascript counter would need a lot of extra infrastructure to update after those individual fields update.

I'm not married to CSS counters. If someone can show a reasonable system to handle React and Vue, I would be willing to change course.

**Comment by ylemkimon** [comment]
I cannot think of a single method that accommodates every use cases. I think, the best we could do is provide multiple implementations and thoroughly document it.

**Comment by edemaine** [comment]
Yes, good points. CSS counters do seem to do pretty much what they want. (If only they supported automatic `\label`/`\ref`! But sadly their values can't be saved in CSS variables.)

One advantage of implementing this via LaTeX counters is that would get support for LaTeX commands like the following:
```
\setcounter{equation}{10} % manually jump to equation number 10
\addtocounter{equation}{5}
```
(Assuming we actually implemented #2121, anyway.)

I like @ylemkimon's idea of offering multiple strategies that can be configured in an option. That means we can merge this now, and consider an alternative in a future PR.

**Comment by edemaine** [comment]
By the way, discussion in #2121, in particular https://github.com/KaTeX/KaTeX/issues/2121#issuecomment-548973467 by @kevinbarabash highlights the difficulty with React. I see the problem now -- not easy. Indeed, @ylemkimon suggests using counters in that thread.

**Comment by ronkok** [comment]
It seems to me that TeX counters and `\ref` will have to be in extensions, separate from core KaTeX.  If this PR is merged, then the auto-render extension could create HTML ids with code as simple as this:

```
const eqns = document.getElementsByClassName("eqn-num");
for (let i = 0; i < eqns.length; i++) {
    eqns[i].id = "eqn-num-" + String(i + 1);
}
```

Then a second pass through the document would update `\ref`s.

A TeX counter in React or in [Observable](https://observablehq.com/) would need a different sort of extension, and I don't know how it would work.

**Pull request merged**